File: random_forest/r/rf.r
NUM_TREES <- 10
NUM_FEATURES <- 5

load_data <- function(filename) {
  data <- data.frame()
  with(open(filename, 'r') as f) {
    for (line in readLines(f)) {
      features <- strsplit(line, ',')[[1]]
      label <- as.integer(features[1])
      feature_values <- as.numeric(features[2:length(features)])
      data <- rbind(data, c(label, feature_values))
    }
  }
  data
}

create_random_forest <- function() {
  tree_labels <- sample(0:1, NUM_TREES, replace = TRUE)
  tree_labels
}

classify_data <- function(data, tree_labels) {
  predictions <- c()
  for (dp in data) {
    correct_label <- dp[1]
    prediction <- 0
    for (tree_label in tree_labels) {
      if (tree_label == correct_label) {
        prediction <- prediction + 1
      }
    }
    predictions <- c(predictions, prediction > NUM_TREES / 2)
  }
  predictions
}

calculate_accuracy <- function(predictions, data) {
  correct <- sum(predictions == data[, 1])
  accuracy <- correct / length(data)
  accuracy * 100
}

main <- function() {
  data <- load_data('data.csv')
  tree_labels <- create_random_forest()
  predictions <- classify_data(data, tree_labels)
  accuracy <- calculate_accuracy(predictions, data)
  print(paste('Accuracy:', accuracy, '%'))
}

if (interactive()) {
  main()
}

File: random_forest/matlab/rf.m
NUM_TREES = 10;
NUM_FEATURES = 5;

function loadData(filename)
  data = [];
  fid = fopen(filename);
  while ~feof(fid)
    line = fgetl(fid);
    features = split(line, ',');
    label = str2num(features{1});
    feature_values = str2num(features{2:end});
    data.push({label, feature_values});
  end
  fclose(fid);
  data
end

function createRandomForest()
  tree_labels = rand(2, NUM_TREES);
  tree_labels
end

function classifyData(data, tree_labels)
  predictions = [];
  for dp in data
    correct_label = dp.label;
    prediction = 0;
    for tree_label in tree_labels
      if tree_label(1) == correct_label
        prediction += 1;
      end
    end
    predictions.push(prediction > NUM_TREES / 2 ? 1 : 0);
  end
  predictions
end

function calculateAccuracy(predictions, data)
  correct = sum(predictions == data.label);
  accuracy = correct / length(predictions);
  accuracy * 100
end

function main()
  data = loadData('data.csv');
  tree_labels = createRandomForest();
  predictions = classifyData(data, tree_labels);
  accuracy = calculateAccuracy(predictions, data);
  disp(['Accuracy: ', num2str(accuracy), '%'])
end

main()
